primitives/avn-common/src/event_discovery.rs

```diff
@@ -90,29 +108,28 @@ pub mod events_helpers {
     pub extern crate alloc;
     use alloc::collections::BTreeSet;
 
-    pub fn discovered_eth_events_partition_factory(
+    pub fn discovered_eth_events_partition_factory(range: EthBlockRange,
         events: Vec<DiscoveredEvent>,
-    ) -> Vec<DiscoveredEthEventsFraction> {
+    ) -> Vec<EthereumEventsPartition> {
         let mut sorted = events.clone();
         sorted.sort();
         let chunk_size: usize = <EventsBatchLimit as sp_core::Get<u32>>::get() as usize;
-        let mut fractions = Vec::<DiscoveredEthEventsFraction>::new();
+        let mut partitions = Vec::<EthereumEventsPartition>::new();
 
         let mut iter = sorted.chunks(chunk_size).enumerate();
-        let fraction_count = sorted.chunks(chunk_size).count() as u16;
-        let hash: H256 = blake2_256(&(&events, fraction_count).encode()).into();
+        let partitions_count = sorted.chunks(chunk_size).count();
 
-        let _ = iter.try_for_each(|(fraction, chunk)| -> Result<(), ()> {
+        let _ = iter.try_for_each(|(partition, chunk)| -> Result<(), ()> {
             let inner_data: BTreeSet<DiscoveredEvent> = chunk.iter().cloned().collect();
             let data = EthEventsPartition::try_from(inner_data)?;
-            fractions.push(DiscoveredEthEventsFraction::new(
+            partitions.push(EthereumEventsPartition::new(
+                range.clone(),
+                partition as u16,
+                partitions_count == partition.saturating_add(1),
```

Comment:
does this `saturating_add` actually change the value of `partition` or not? This is relevant, because that would be introducing a side-effect (which is discouraged in functional programming, because in general it makes validating code correctness harder) that is beyond what we need in here.
I don't know exactly how saturating methods work, since they have an implicit argument (the self) and the look of them _could mean either implementation_, but the look of it suggests it changes the callee.
So, I'm not comfortable with using them unless we absolutely need to.

Reply:
https://docs.rs/sp-runtime/latest/sp_runtime/traits/trait.Saturating.html#tymethod.saturating_add
self is immutable, no implementation can change `self` values.
See https://docs.rs/sp-runtime/latest/sp_runtime/traits/trait.Saturating.html#method.saturating_inc that does.
I can make the implementation use an immutable iterator to the partitions if that will give you more confidence.